### PR TITLE
feat: add MiniCPM4 Eagle speculative decoding

### PR DESCRIPTION
**Pull request by wooway777**
```
python examples/test_infer.py   --model=/data-aisoft/mechdancer/models/openbmb_MiniCPM4-8B/   --draft-model=/data-aisoft/mechdancer/models/openbmb_MiniCPM4-8B-Eagle-vLLM/   --enable-paged-attn --attn=flash-attn --num-draft-tokens=4
```
image

```
python examples/test_infer.py   --model=/data-aisoft/mechdancer/models/openbmb_MiniCPM4-8B/    --enable-paged-attn --attn=flash-attn
```
image

Qwen 3
image

MiniCPM V
image

VideoNSA
f5a96ff1-c807-463f-b798-652eca672fd3

### CODE REVIEW

File: python/infinilm/infer_engine.py [hunk 1]
```diff
@@ -241,6 +265,99 @@ def convert_tensor_list(tensor_list_):
             handle_oom_and_exit(e)
             raise
 
+    def forward_raw(
```
Comment (pengcheng888) [inline python/infinilm/infer_engine.py:399]: 只有forward_raw才有  visual_token_ranges, 是这样么

Comment (pengcheng888) [reply python/infinilm/infer_engine.py:399]: forward_raw 和 forward两个函数代码重复太高了

File: csrc/pybind11/engine/engine.hpp [hunk 2]
```diff
@@ -109,14 +109,16 @@ inline void bind_infer_engine(py::module &m) {
         .def("process_weights_after_loading", &InferEngine::process_weights_after_loading, "Process the weights after loading on all workers (e.g., for quantization)")
         .def(
             "forward", [](InferEngine &self, const InferEngine::Input &input) -> InferEngine::Output {
-                // IMPORTANT: Release the GIL before calling forward() to allow other Python threads
-                // to run concurrently during inference (which may block for a long time).
-                // Do NOT remove this — without it, the GIL is held throughout inference and will
-                // deadlock or stall any other Python thread (e.g., request handling, scheduling).
```
Comment (pengcheng888) [inline csrc/pybind11/engine/engine.hpp:119]: 这四行注释是马老是特意加上了，一是提示，二是不让format改来改去。 一不小心把 py::gil_scoped_release release 改没了。  

还有保留注释，可以修改为比较规范的注释形式， 这样ai可能就不再删除了

File: csrc/models/minicpm4/minicpm4_allocate_kv_cache_tensors.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace infinilm::models::minicpm4 {
+
+std::vector<infinicore::Tensor> minicpm4_allocate_kv_cache_tensors(const cache::CacheConfig *cache_config,
+                                                                   const std::shared_ptr<infinilm::config::ModelConfig> &text_config,
+                                                                   const backends::AttentionBackend &attention_backend) {
+    if (nullptr == cache_config) {
```
Comment (pengcheng888) [inline csrc/models/minicpm4/minicpm4_allocate_kv_cache_tensors.cpp]: minicpm4_allocate_kv_cache_tensors 函数里的代码，看着跟默认实现的那个函数没有什么差别，

这个minicpm4_allocate_kv_cache_tensors函数是不是可以删掉。

File: csrc/models/minicpm_eagle/minicpm_eagle_allocate_kv_cache_tensors.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
+#include "../../cache/kv_cache.hpp"
+#include "minicpm_eagle_for_causal_lm.hpp"
+
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace infinilm::models::minicpm_eagle {
+
+std::vector<infinicore::Tensor> minicpm_eagle_allocate_kv_cache_tensors(const cache::CacheConfig *cache_config,
```
Comment (pengcheng888) [inline csrc/models/minicpm_eagle/minicpm_eagle_allocate_kv_cache_tensors.cpp]: 同理，可以删除么minicpm_eagle_allocate_kv_cache_tensors

File: csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+#include "../../utils.hpp"
+#include "../models_registry.hpp"
+
+#include "infinicore/ops.hpp"
+
+#include <cmath>
+#include <stdexcept>
+#include <string>
+
+namespace infinilm::models::minicpm_eagle {
```
Comment (pengcheng888) [inline csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp:10]: minicpm_eagle和minicpm除了residual_scale函数的返回值外，Attention MLP  DecoderLayer三个模块看着都是一样么。可以直接minicpm_eagle复用minicpm的么，而不是重新定义一遍。

File: csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    INFINICORE_NN_MODULE_INIT(model, model_config, device);
+    INFINICORE_NN_MODULE_INIT(lm_head, hidden_size, vocab_size, false, dtype, device);
+
+    if (model_config->get_config_json().contains("dim_model_base")) {
+        const float dim_model_base = model_config->get<float>("dim_model_base");
+        lm_head_->set_alpha(dim_model_base / static_cast<float>(hidden_size));
+    }
+}
+
+infinilm::InfinilmModel::Output MiniCPMEagleForCausalLM::forward(const infinilm::InfinilmModel::Input &input) const {
```
Comment (pengcheng888) [inline csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp:83]: 为什么在MiniCPMEagleForCausalLM中没有 看到mtp相关的权重

File: python/infinilm/processors/sentencepiece_processor.py [hunk 7]
```diff
@@ -75,6 +75,12 @@ class MistralProcessor(SentencePieceProcessor):
     pass
 
 
+@register_processor("minicpm")
```
Comment (pengcheng888) [inline python/infinilm/processors/sentencepiece_processor.py:79]: visual_token_ranges是添加videonsa模型才新增的meta类型的数据，逻辑在vedionsa_processer.py文件中。

 minicpm和minicpm4的Processor中有处理visual_token_ranges的数据的逻辑么

Comment (pengcheng888) [reply python/infinilm/processors/sentencepiece_processor.py:79]: 该pr中visual_token_ranges在哪里被赋值的

File: python/infinilm/llm/model_runner/speculative_runner.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
+import infinicore
+
+from infinilm.cache.cache import StaticKVCacheConfig
+from infinilm.distributed import DistConfig
+from infinilm.infer_engine import InferEngine
+from infinilm.modeling_utils import load_model_state_dict_by_file
+
+
+class SpeculativeRunner:
```
Comment (pengcheng888) [inline python/infinilm/llm/model_runner/speculative_runner.py:8]: SpeculativeRunner和Runner这两个类的关系和代码层级，需要重新梳理一下，再改。

File: python/infinilm/infer_engine.py (hunk 1 above)
Comment (wooway777) [reply python/infinilm/infer_engine.py:399]: 这个是因为rank worker依赖forward_raw函数名去做run_raw的操作

File: csrc/pybind11/engine/engine.hpp (hunk 2 above)
Comment (wooway777) [reply csrc/pybind11/engine/engine.hpp:119]: 我把这四行评论恢复了这里还是显示删除

File: python/infinilm/processors/sentencepiece_processor.py (hunk 7 above)
Comment (wooway777) [reply python/infinilm/processors/sentencepiece_processor.py:79]: 应该不涉及visual的问题，这里声明的事minicpm模型，不是minicpmv模型。目前用到的两个model type都是写的minicpm，也许不该加那个minicpm4？

File: csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp (hunk 6 above)
Comment (wooway777) [reply csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp:83]: 真正 Eagle/MTP 相关权重在 MiniCPMEagleModel 里面，不是在 MiniCPMEagleForCausalLM 外层直接展开：
embed_tokens
input_norm1
input_norm2
fc
eagle_layers
norm
lm_head
其中比较关键的 MTP/Eagle 特有部分是：
input_norm1
input_norm2
fc
它们对应 Eagle draft 的输入融合路径：把 draft token embedding 和 target/base model hidden states 拼起来，然后通过 fc 投影回 hidden size：
auto input_embeds = input_norm1_->forward(embed_input_ids(input_ids));
auto target_hidden = input_norm2_->forward(target_hidden_states);

auto fused_input = Tensor::empty(... hidden_size * 2 ...);
fused_input[..., :hidden_size] = input_embeds;
fused_input[..., hidden_size:] = target_hidden;

auto hidden_states = fc_->forward(fused_input);
也就是说，MTP/Eagle 的核心不是一个单独叫 mtp_* 的权重，而是这组 “token embedding + target hidden fusion” 权重，以及后面的 eagle_layers 和 lm_head。

File: csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp (hunk 5 above)
Comment (wooway777) [reply csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp:10]: 删掉了 MiniCPMEagleAttention
删掉了 MiniCPMEagleMLP
删掉了 MiniCPMEagleDecoderLayer
MiniCPMEagleModel 里现在直接用：
INFINICORE_NN_MODULE_VEC(
    infinilm::models::minicpm4::MiniCPM4DecoderLayer,
    eagle_layers
);

Comment (pengcheng888) [reply csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp:10]: using MiniCPMegare4DecoderLayer =   infinilm::models::minicpm4::MiniCPM4DecoderLayer,

再MiniCPMegare中使用MiniCPMegare4DecoderLayer，  而非infinilm::models::minicpm4::MiniCPM4DecoderLayer。

File: csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp (hunk 6 above)
Comment (pengcheng888) [reply csrc/models/minicpm_eagle/minicpm_eagle_for_causal_lm.cpp:83]: 我再理解理解

File: python/infinilm/llm/cache_manager.py [hunk 9]
```diff
@@ -337,6 +339,60 @@ def allocate_slots(
...
+        total_token_ids: List[int] = None,
+    ) -> tuple[List[int], List[int]]:
+        """Append multiple decode slots for speculative target verification."""
+        slots = []
+        for offset in range(num_slots):
+            block_table, slot = self.append_slot(
+                block_table, start_num_tokens + offset, total_token_ids
+            )
+            slots.append(slot)
+        return block_table, slots
```
Comment (ma-hang) [inline python/infinilm/llm/cache_manager.py:397]: 预测token所在block的hash也可能会被更新，进而被同batch的后续请求复用，而这些token还不确定会被接受

Comment (wooway777) [reply python/infinilm/llm/cache_manager.py:397]: 当追加 token 后跨到新 block 时，说明前一个 block 已经满了。普通 decode 下，这个 full block 可以进入 prefix cache：
self.hash_to_block_id[current_hash] = last_block_id
但是 speculative verify 阶段不一样。它会临时把 draft tokens 追加进 KV cache，用 target 一次 verify 多个 token。此时这些 token 还没最终确认，后面可能被 reject，也可能 correction。review 说的就是这个问题：如果这个临时 full block 立刻更新 hash，同 batch 后续请求可能复用到“还没最终确认甚至可能会被改掉”的预测 token KV。
所以我没有新写一套 append 逻辑，而是在原来的 append_slot 上加了一个开关：
update_hash: bool = True
默认还是 True，普通 decode 行为不变。
然后 append_slots 也透传这个开关：
append_slots(..., update_hash=False)
SpeculativeRunner 里现在这样调用：
cache_manager.append_slots(
...,
update_hash=False,
)
这样做的原因是：speculative verify 用的批量追加最终还是通过 append_slot 跨 block 分配 slot；如果不改 append_slot，它内部仍然会在跨 block 时更新 hash，污染 prefix cache。
改原来的函数是为了让这条共享底层逻辑能“分配 slot 但不注册 prefix hash”。
普通推理还是：
append_slot(..., update_hash=True) # 默认
所以其他模型和普通服务路径不受影响。

File: python/infinilm/llm/cache_manager.py [hunk 10]
```diff
@@ -337,6 +339,60 @@ def allocate_slots(
...
+
+        removed = block_table[keep_blocks:]
+        for block_id in removed:
+            block = self.blocks[block_id]
+            if block.hash != -1 and self.hash_to_block_id.get(block.hash) == block_id:
+                del self.hash_to_block_id[block.hash]
+            block.hash = -1
+            block.token_ids = []
+            if block.ref_count > 0:
+                block.ref_count -= 1
```
Comment (ma-hang) [inline python/infinilm/llm/cache_manager.py]: 这里ref_count应该直接置零，block已经失效；
这里以及block manager的其他地方最好根据需求使用封装好的block管理函数，如：_deallocate_block、free_blocks等

File: python/infinilm/llm/llm.py [hunk 11]
```diff
@@ -170,57 +170,63 @@ def _update_requests(
...
 
-            if not holds_back:
-                req.generated_text = last_committed_text + delta
-                req._token_decode_offset = len(req.generated_token_ids)
+                last_committed_text = req.generated_text
 
-            is_finished = self._check_request_finished(req, token_id)
+                if not holds_back:
+                    req.generated_text = last_committed_text + delta
+                    req._token_decode_offset = len(req.generated_token_ids)
```
Comment (ma-hang) [inline python/infinilm/llm/llm.py:206]: last_committed_text = req.generated_text为什么移到 if not holds_back外，确认逻辑正确

File: python/infinilm/llm/scheduler.py [hunk 12]
```diff
@@ -19,10 +19,12 @@ def __init__(
         self,
         scheduled_requests: List[InferenceRequest],
         is_prefill: bool = False,
+        cache_manager: BlockManager | None = None,
     ):
         self.scheduled_requests = scheduled_requests
         self.num_requests = len(scheduled_requests)
         self.is_prefill = is_prefill
+        self.cache_manager = cache_manager
```
Comment (ma-hang) [inline python/infinilm/llm/scheduler.py]: cache_manager传给model runner是要做什么处理，非必要不应该传输

**Review by ma-hang — COMMENTED** [review]
需要修改

File: python/infinilm/infer_engine.py (hunk 1 above)
Comment (wooway777) [reply python/infinilm/infer_engine.py:399]: 改了一版不用forward_raw的

File: python/infinilm/llm/llm.py (hunk 11 above)
Comment (wooway777) [reply python/infinilm/llm/llm.py:206]: 已经按评论改到 if not holds_back 内，逻辑正确

File: python/infinilm/llm/scheduler.py (hunk 12 above)
Comment (wooway777) [reply python/infinilm/llm/scheduler.py]: cache_manager 传给 model runner：已经改成 SpeculativeCacheOps，不再透完整 BlockManager

File: python/infinilm/llm/cache_manager.py (hunk 9 above)
Comment (wooway777) [reply python/infinilm/llm/cache_manager.py:397]: 代码已通过 update_hash=False 和 SpeculativeCacheOps.append_verify_slots() 处理

File: python/infinilm/llm/llm.py (hunk 11 above)
Comment (wooway777) [reply python/infinilm/llm/llm.py:206]: image

File: python/infinilm/llm/cache_manager.py [hunk 13]
```diff
@@ -337,8 +339,63 @@ def allocate_slots(
 
         return block_table, slot_mapping
 
+    def append_slots(
```
Comment (wooway777) [inline python/infinilm/llm/cache_manager.py:379]: append_slots() 是给 speculative verify 用的“批量追加 decode slots”辅助函数。
原来的 append_slot() 一次只给 1 个新 token 分配 KV slot。普通 decode 每轮只生成 1 个 token，所以够用。
但 Eagle/MTP speculative decoding 里流程是：
base model 先产出当前轮 target token 和 hidden state。
Eagle draft model 一次草拟多个 token，比如 num_draft_tokens=4。
如果第 1 个 draft token 和 target token 对上，就需要让 target model 一次性 verify 后面多个 draft token。
这时 target model 的输入包含多个 token，所以 KV cache 也必须一次性预留多个 slot。

从 start_num_tokens 开始，连续追加 num_slots 个 KV slot。
必要时分配新的 KV block。
返回更新后的 block_table。
返回每个新 token 对应的 slot_mapping，供 target verify forward 使用。
透传 update_hash，让普通追加保持默认 hash 行为，而 speculative verify 可以用 update_hash=False 避免未确认 draft token 污染 prefix cache。

只有投机推理会用到append_slots

**Comment by wooway777** [comment]
image
@ma-hang
